File: src/main/java/edu/kit/ifv/mobitopp/actitopp/HTrip.java
package edu.kit.ifv.mobitopp.actitopp;


/**
 * 
 * @author Tim Hilgert
 *
 */
public class HTrip {

	// each trip is bound to an activity
	private HActivity activity;
	
	// indicator if trip is one before or after an activity
	private TripStatus status;

	private int tripduration = -1;


	public HTrip(HActivity parent, TripStatus type, int tripduration)
	{
		assert parent!=null :	"corresponding activity NULL!";
		assert tripduration>0 :	"duration is less or equal 0!";
		
		this.activity = parent;
		this.status = type;
		this.tripduration = tripduration;
	}

	/**
	 * @return the tripduration
	 */
	public int getDuration() {
		return tripduration;
	}


	/**
	 * @param tripduration the tripduration to set
	 */
	public void setDuration(int tripduration) {
		this.tripduration = tripduration;
	}

	/**
	 * @return the activity
	 */
	public HActivity getActivity() {
		return activity;
	}
	
	/**
	 * 
	 * @return
	 */
	public boolean isScheduled()
	{
		return getDuration()!=-1;
	}
	
	/**
	 * 
	 * @return
	 */
	public int getStartTime()
	{
		int starttime=-1;
		
		if (status.equals(TripStatus.TRIP_BEFORE_ACT))
		{
			starttime = activity.getStartTime() - tripduration;
		}
		if (status.equals(TripStatus.TRIP_AFTER_ACT))
		{
			starttime = activity.getEndTime();
		}
		
		assert starttime!=-1 : "could not get TripStartTime";		
		return starttime;
	}
	
	/**
	 * 
	 * @return
	 */
	public int getEndTime()
	{
		int endtime=-1;
		
		if (status.equals(TripStatus.TRIP_BEFORE_ACT))
		{
			endtime = activity.getStartTime();
		}
		if (status.equals(TripStatus.TRIP_AFTER_ACT))
		{
			endtime = activity.getEndTime() + tripduration;
		}
		
		assert endtime!=-1 : "could not get TripEndTime";		
		return endtime;
	}
	
	/**
	 * 
	 * @return
	 */
	public int getStartTimeWeekContext()
	{
		int starttime=-1;
		
		if (status.equals(TripStatus.TRIP_BEFORE_ACT))
		{
			starttime = activity.getStartTimeWeekContext() - tripduration;
		}
		if (status.equals(TripStatus.TRIP_AFTER_ACT))
		{
			starttime = activity.getEndTimeWeekContext();
		}
		
		assert starttime!=-1 : "could not get TripStartTime";		
		return starttime;
	}
	
	/**
	 * 
	 * @return
	 */
	public int getEndTimeWeekContext()
	{
		int endtime=-1;
		
		if (status.equals(TripStatus.TRIP_BEFORE_ACT))
		{
			endtime = activity.getStartTimeWeekContext();
		}
		if (status.equals(TripStatus.TRIP_AFTER_ACT))
		{
			endtime = activity.getEndTimeWeekContext() + tripduration;
		}
		
		assert endtime!=-1 : "could not get TripEndTime";		
		return endtime;
	}
	
	/**
	 * 
	 * @return
	 */
	public ActivityType getType()
	{
		ActivityType type=ActivityType.UNKNOWN;
		
		if (status.equals(TripStatus.TRIP_BEFORE_ACT))
		{
			type = activity.getActivityType();
		}
		
		/*
		 * until now, trip after activities only occur after the last activity in a tour, thus they are always trips to home
		 */
		if (status.equals(TripStatus.TRIP_AFTER_ACT))
		{
			type = ActivityType.HOME;
		}
		
		assert type!=ActivityType.UNKNOWN : "could not get TripType";		
		return type;
	}
	
	/**
	 * 
	 * @return
	 */
	public JointStatus getJointStatus()
	{
		JointStatus jointStatus= JointStatus.UNKNOWN;
		
		if (status.equals(TripStatus.TRIP_BEFORE_ACT))
		{
			jointStatus = activity.getJointStatus();
		}
		
		/*
		 * until now, trip after activities are home trips and joint home trips are not yet supported
		 * 
		 */
		if (status.equals(TripStatus.TRIP_AFTER_ACT))
		{
			jointStatus = JointStatus.NOJOINTELEMENT;
		}
		
		assert jointStatus!=JointStatus.UNKNOWN : "could not get jointStatus";		
		return jointStatus;
	}
	
	
	@Override
	public String toString()
	{
		return "trip: start(week): " + getStartTimeWeekContext() + " / end(week): " + getEndTimeWeekContext() + " / duration: " + getDuration();
	}
	
}

File: src/main/java/edu/kit/ifv/mobitopp/actitopp/HTrip.kt
package edu.kit.ifv.mobitopp.actitopp

import edu.kit.ifv.mobitopp.actitopp.enums.ActivityType
import edu.kit.ifv.mobitopp.actitopp.enums.JointStatus
import edu.kit.ifv.mobitopp.actitopp.enums.TripStatus

/**
 * A Trip in actitopp has:
 * @property activity an activity from which the trip is spawned
 * @property duration a duration encoded in a naked int. Presumed unit is minutes. TODO make duration instead.
 */
interface Trip {
    val activity: HActivity
    var duration: Int
}
/**
 * @author Tim Hilgert
 */
class HTrip(parent: HActivity, type: TripStatus, tripduration: Int) {

    init {
        require(tripduration != -1) {
            "This require block tries to locate scenarios where a trip is set to -1 duration, which has a specific encoding, should this happen you know" +
                    "that somewhere in actiTopp an unspecified duration is required, If you never see this message, you can kill the [isScheduled] field"
        }
    }

    var duration: Int = tripduration
        set(value) {
            require(value != -1) {
                "This require block tries to locate scenarios where a trip is set to -1 duration, which has a specific encoding, should this happen you know" +
                        "that somewhere in actiTopp an unspecified duration is required, If you never see this message, you can kill the [isScheduled] field"
            }
            field = value
        }
    /**
     * @return the activity
     */
    // each trip is bound to an activity
    val activity: HActivity = parent

    // indicator if trip is one before or after an activity
    private val status: TripStatus = type
    // The setter is only called in HActivity.createTripsforActivity()


    val isScheduled: Boolean get()  {

        return duration != -1
    }

    val startTime: Int get() {
        return when(status) {
            TripStatus.TRIP_BEFORE_ACT -> activity.startTime - duration
            TripStatus.TRIP_AFTER_ACT -> activity.endTime
        }
    }

    val endTime: Int get() {
        return when(status) {
            TripStatus.TRIP_BEFORE_ACT -> activity.startTime
            TripStatus.TRIP_AFTER_ACT -> activity.endTime + duration
        }
    }
    val startTimeWeekContext: Int get()  {
        return when(status) {
            TripStatus.TRIP_BEFORE_ACT -> activity.startTimeWeekContext - duration
             TripStatus.TRIP_AFTER_ACT -> activity.endTimeWeekContext
        }
    }
    val endTimeWeekContext: Int get() {
        return when(status) {
            TripStatus.TRIP_BEFORE_ACT -> activity.startTimeWeekContext
            TripStatus.TRIP_AFTER_ACT -> activity.endTimeWeekContext + duration
        }
    }
    // Apparently TRIP_AFTER_ACT means going Home  - Robin, 25
    val type: ActivityType get() {
        return when(status) {
            TripStatus.TRIP_BEFORE_ACT -> activity.activityType
            TripStatus.TRIP_AFTER_ACT -> ActivityType.HOME
        }
    }
    val jointStatus: JointStatus get() {
        return when(status) {
            TripStatus.TRIP_BEFORE_ACT -> activity.jointStatus
            TripStatus.TRIP_AFTER_ACT -> JointStatus.NOJOINTELEMENT // Apparently the original implementation does not support home trips, that cannot be joint home trips, for some reason
            //TODO debug and find out why in the legacy code the comment said "Home trips cannot be joint trips YET" - Robin, 25
        }
    }


    override fun toString(): String {
        return "trip: start(week): $startTimeWeekContext / end(week): $endTimeWeekContext / duration: $duration"
    }
}
